### ISSUE TITLE
Convert all unicode to str before calling python command scripts?

### ISSUE THREAD

**Issue by ywangd** [ywangd/stash#62]
Because Pythonista (or iOS) UIs by default use unicode, all the inputs read from the `TextField` is of unicode type. In the meanwhile, Python 2.x is not fully unicode compatible. Therefore a rather elusive bug is generated when a Python script wants strictly a `str` but an `unicode` is provided. Although values of the two types are essentially the same, e.g. `'SimpleHTTPServer'` v.s. `u'SimpleHTTPServer'`, the script sees the type difference and errors out. 

This issue has been encountered twice by @briarfox with the `python` command and @jsbain with the `git` command (#57).

As @jsbain suggested, this bug breaks the paradigm that generic python scripts could run in stash without changes. So this time I am really thinking about converting all the `unicode` inputs to `str` type before passing them to external command scripts. @dgelessus 

Your comments and suggestions are highly appreciated.

**Comment by dgelessus** [comment]
Python 2 has very fuzzy Unicode handling indeed. The issue is that `str` containing UTF-8 and `unicode` might seem compatible and interchangeable (because `"string" + u"ünicode" == u"stringünicode"`) but they are actually separate things. Python 2 (unlike Python 3) coerces `str` and `unicode` where it can, except when working with Python-internal identifiers, such as with `getattr(obj, "foo")` or `globals()["name"]`, both of which only work with `str`, not `unicode`.

I'm not very familiar with the Stash internals, but it looks like they rely on this fuzzy implicit `str`/`unicode` conversion. Please don't read this as an accusation/"you're doing it wrong", it is understandable to think that normal unprefixed `"strings"` are what should be used in most cases, but that isn't true. In general anything that is human-readable text should be a Unicode string - user command input, button labels, console output, etc. The only exceptions are Python identifiers, which need to be `str` because we're using Python 2.

When working with raw binary data it may also be helpful to specifically use the `bytes` name and `b""` literal syntax. In Python 2 these are equivalent to `str` and `""`, but by using the Python 3-like naming it is always clear what the string is meant to be - `u""` or `unicode` is text, `b""` or `bytes` is raw data, and `""` or `str` is a Python identifier. If Pythonista should ever be updated to use Python 3 this would also make updating the code easier.

**Comment by jsbain** [comment]
The underlying question should be what format should user input be?  Pythonista's stdin uses `str`, as does the long press play to set arguments, as does arguments coming from pythonista:// Uris.  Shellista, using stdin, also gives str.   So, I'd argue stash should as well.  I have a hard time thinking of a case where unicode storage is important, given that UTF8 str can encode all unicode characters.   The one case might be writing to a unicode file.  Maybe if any network protocols expect unicode, but I don't know of any.  Most real terminals that I'm aware of use UTF8 or ascii.

Given most commands either don't care, or specifically expect str, the default ought to be str, not unicode, so that the methods that work at the console should work in stash as well.

It is a little bizarre thinking of filenames or commands named 👩💧🐭🏯, but UTF8 would support that, we don't need to use unicode types.  If any commands want to specifically written in unicode, which are few I think, those functions should cast to unicode, just like the would need to from the console.

**Comment by dgelessus** [comment]
@jsbain, I think you may be misunderstanding the relationship between Unicode and UTF-8. Unicode is a standard that assigns glyphs (ä) to code points (U+00E4), and UTF-8 is an encoding that allows for more efficient and portable storage of Unicode text than storing each character as a 16-bit (or more) integer. Because of that you'll rarely ever find raw Unicode data in a file (technically that would be UTF-16 or UTF-32 anyway).

The reason why using Python's `unicode` type is better than using UTF-8 bytes in a `str` is because `unicode` is Python's standard Unicode type. Using UTF-8 `str`s instead of `unicode` is like using four-element `bytearray`s to store 32-bit numbers. `unicode` stores characters unambiguously, unlike `str`, where you need to worry whether it's UTF-8/16/32, ISO-8859-[1-13], or some Windows codepage.

Half of this `str` vs. `unicode` issue is because we're using Python 2 anyway, which has to remain backwards-compatible with old versions that didn't have any built-in Unicode support. In Python 3 the two types are differentiated much more clearly - `str` has full Unicode support, `bytes` is an immutable byte sequence like Python 2 `str`, and there is no coercion or implicit en/decoding. When using Python 3 you probably wouldn't be asking, "should I use `bytes` or `str` to store text?", the answer would be obvious. Now think about Python 2, where `str` is Py3 `bytes` and `unicode` is Py3 `str`. Again the choice should be clear.

**Comment by ywangd** [comment]
@jsbain @dgelessus Thanks a lot for your valuable inputs. They also help me understand the issue better. 

Taking account of all discussions, I am going to make following changes:

* StaSh **by default** casts all inputs received by TextField from `unicode` type to `str` type with utf-8 encoding. 
* An option will be available (through the config file `.stash_config`) to turn off the default behaviour and leave unicode inputs untouched.

The biggest reason for the default type cast is to be consistent with most other input methods available in Pythonista, e.g. interactive prompt, and even consistent with PC version Python prompt. The changes are in line with the principle of least surprise to avoid unnecessary confusions in future developments. 

I do agree with @dgelessus that unicode is in theory an ultimate better choice and this issue is largely due to Python 2's incomplete unicode support. So when Pythonista is upgraded with Python 3 support, this issue is most likely to go away by itself. But before that happens (I am not aware of any timeline), I think it is better to stick with `str` which is what most existing Python 2 scripts expect.

**Comment by dgelessus** [comment]
FYI, I have a branch in my fork where I'm currently working on unicodifying the StaSh core and eventually the commands. I'm only switching those strings to `unicode` that are not script-internal, i. e. exception messages and internal parser strings are still normal `str`.

(I still need to actually commit the changes to `stash.py`, in its current local state it doesn't work very well yet.)

**Comment by ywangd** [comment]
The changes have been made so by default stash convert all unicode inputs to str type with utf-8 encoding. This happens in two places:

1. Right before the arguments are passed to external scripts. The arguments are cast to `str`
2. When the read related methods (`read`, `readline`, `readlines`) are called, the input buffer string will be converted to `str`.

An configuration option `input_encoding_utf8=1` is set to enable the default encoding behavior. 20198fce76d524a046c7e2b947bd26b90f592de9

@dgelessus I kinda think the opposite way. My preference is that StaSh internals (mainly the parser bit) use `unicode` and conversion only happens when an external scripts asking for inputs/arguments. In another word, the external I/O interface to Stash is by default `str` type, while internals are`unicode`. A few reasons are:

* Though Stash in its current state actually cannot parse `unicode` input, it seems possible to make pyparsing work with `unicode`. In addition, it seems pyparsing has difficulties working with utf-8 encoded `str` (strings like `'\xd3\x92'` makes pyparsing complain)
* It will be easy to upgrade to Python 3 when it is out since all internals are `unicode` anyway.
* By flipping a switch, external I/O interfaces will also use `unicode`, this also helps future upgrade.

Honestly I am no expert in this unicode business. It seems to have much deeper resonance than I had expected (not just speaking of Python) and I am still getting the hang of it. Thanks again for your valuable inputs.

**Comment by dgelessus** [comment]
Interaction between scripts/commands needs to be done using `str` no matter what, because the stdin/out/err streams are byte-based in Python 2 and 3 (the latter uses UTF-8 as the default encoding) and `sys.argv` uses the native `str` type for the respective version. (I'm not sure what exactly you mean by "external I/O". The I/O streams? File I/O? Runtime args? Interaction wih `ui`?)

**Comment by ywangd** [comment]
@dgelessus Sorry about the confusion. Please let me rephrase.

By "external I/O", I mean interactions between stash and command scripts, i.e. bin commands. There are mainly two types of interactions:
 1. Stash passes arguments to a command script
 2. A command script is asking for input via stash's TextField input.

An example for the first type interaction. When stash receives input `python -m SimpleHTTPServer`. It is an `unicode` string. So in Python language, the input is `u'python.py -m SimpleHTTPServer'`. This unicode string will be processed internally (e.g. parsing) as is by Stash. When it is about to invoke the command `python.py`, the arguments, i.e. `u'-m'` and `u'SimpleHTTPServer'`, are converted to `str` type, i.e. `'-m'` and `'SimpleHTTPServer'`, so that the `python.py` script only sees `str` type arguments. This conversion eliminates the bugs that @briarfox  and @jsbain had for the `python` and `git` commands.

The second type interactions is when an running command asks for input from user. For an example, when the `cat.py` command is invoked without arguments, it asks for input from users, i.e. it requires users typing into the Stash's TextField. The input strings received from the TextField are again unicode and they are converted to `str` type before passing to `cat.py`.

Other than the above two cases, I consider everything else as Stash internal (where unicode is preserved).

PS: Apparently another interaction between stash and command script is when  a command sends (prints) output to Stash's terminal. But this seems to be less important since the terminal UI uses `unicode` exclusively and conversion from `str` to `unicode` is automatic. So it does not matter what a command script sends through. It will be displayed as `unicode`.
